fix: do not auto destroy request stream

achingbrain opened the pull request:
Auto destroying the request stream stops `for await..of`-style iteration over request payloads working when `output: 'stream', parse: false` are used as request handler payload config options.

Fixes hapijs/hapi#4149

achingbrain commented:
This is another attempt to resolve the node 14 streaming bug - it's smaller scope than #129 and only addresses the issue I encountered.

The hapi test suite from master passes with this change made to shot.

kanongil commented:
This is not the right fix. The simple fix, that preserves as much as possible, is to just add `{ autoDestroy: false`} to the `super()` call. This effectively reverts [this](https://github.com/nodejs/node/pull/30623) node 14.0 feature for shot.

achingbrain commented:
> just add `{ autoDestroy: false }` to the super() call.

I've added this to the constructor and removed the `destroy` method since the node docs say it shouldn't be overridden.

`autoDestroy` was added in node 10 so it should work for supported versions.

kanongil commented:
Please don't remove the destroy overload for a minimal patch, as it could create extra 'close' emits (from `autoClose: true`), and other unforeseen changes.

achingbrain commented:
I've restored it, though the docs say not to override it.

cjihrig commented:
I think we should go with something more like https://github.com/hapijs/shot/pull/133, but I'd like to keep the test here.

achingbrain commented:
I have no strong preference either way, I'd just like the bug to be squashed!

cjihrig commented:
#133 has a handful of reviews. Can you please revert the changes to `lib/` in this PR so we can keep the test.

achingbrain commented on lib/request.js:
```diff
@@ -13,7 +13,7 @@ exports = module.exports = internals.Request = class extends Stream.Readable {
     constructor(options) {
 
-        super();
+        super({ autoDestroy: false });
```
```suggestion
        super();
```

devinivy commented:
@achingbrain thanks for the contribution and for bearing with all the back-and-forth on this one.  This bug turned out to be a tricky one.

achingbrain commented:
We got there in the end, thanks for all the help.  Will this be released soon?

devinivy commented:
@achingbrain Yeah, I expect it will.  Someone just has to go through the steps of releasing shot and a patch version of hapi.  My guess is that it will be settled before Monday.
